x-pack/metricbeat/module/gcp/metrics/dataproc/metadata.go

```diff
@@ -0,0 +1,208 @@
+	clusterName string
+	machineType string
+
+	User     map[string]string
+	Metadata map[string]string
+	Metrics  interface{}
+	System   interface{}
+}
+
+type metadataCollector struct {
+	projectID         string
+	projectName       string
+	organizationID    string
+	organizationName  string
+	regions           []string
+	collectUserLabels bool
+	opt               []option.ClientOption
+	// NOTE: clusters holds data used for all metrics collected in a given period
+	// this avoids calling the remote endpoint for each metric, which would take a long time overall
+	clusters map[string]*dataproc.Cluster
+	logger   *logp.Logger
+}
+
+// Metadata implements googlecloud.MetadataCollector to the known set of labels from a Dataproc TimeSeries single point of data.
+func (s *metadataCollector) Metadata(ctx context.Context, resp *monitoringpb.TimeSeries) (gcp.MetadataCollectorData, error) {
+	stackdriverLabels := gcp.NewStackdriverMetadataServiceForTimeSeries(resp, s.organizationID, s.organizationName, s.projectName)
+
+	metadataCollectorData, err := stackdriverLabels.Metadata(ctx, resp)
+	if err != nil {
+		return gcp.MetadataCollectorData{}, err
+	}
+
+	if resp.Resource != nil && resp.Resource.Labels != nil {
+		_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudInstanceIDKey, resp.Resource.Labels["cluster_uuid"])
+	}
+
+	_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudInstanceNameKey, resp.Resource.Labels["cluster_name"])
+
+	if s.collectUserLabels {
+		metadata, err := s.instanceMetadata(ctx, s.instanceID(resp), s.instanceRegion(resp))
+		if err != nil {
+			return gcp.MetadataCollectorData{}, err
+		}
+
+		if metadata.machineType != "" {
+			lastIndex := strings.LastIndex(metadata.machineType, "/")
+			_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudMachineTypeKey, metadata.machineType[lastIndex+1:])
+		}
+
+		metadata.Metrics = metadataCollectorData.Labels[gcp.LabelMetrics]
+		metadata.System = metadataCollectorData.Labels[gcp.LabelSystem]
+
+		if metadata.User != nil {
+			metadataCollectorData.Labels[gcp.LabelUser] = metadata.User
+		}
+	}
+
+	return metadataCollectorData, nil
+}
+
+// instanceMetadata returns the labels of an instance
+func (s *metadataCollector) instanceMetadata(ctx context.Context, instanceID, region string) (*dataprocMetadata, error) {
+	cluster, err := s.instance(ctx, instanceID)
+	if err != nil {
+		return nil, fmt.Errorf("error trying to get data from instance '%s' %w", instanceID, err)
+	}
+
+	metadata := &dataprocMetadata{
+		clusterID: instanceID,
+		region:    region,
+	}
+
+	if cluster == nil {
+		s.logger.Debugf("couldn't get instance '%s' call ListInstances API", instanceID)
+		return metadata, nil
+	}
+
+	if cluster.ClusterName != "" {
+		parts := strings.Split(cluster.ClusterName, "/")
+		metadata.clusterName = parts[len(parts)-1]
+	}
+
+	if cluster.Labels != nil {
+		metadata.User = cluster.Labels
+	}
+
+	return metadata, nil
+}
+
+// instance returns data from an instance ID using the cache or making a request
```

Comment:
Am I missing something or it's only checking the cache?

Reply:
Yes, `instance` method only returns an instance (dataproc cluster) by id from the cache.

Reply:
so, perhaps the doc should reflect it?
```suggestion
// instance returns data from an instance ID from the cache.
```